internal/cephfs/fuserecovery.go

```diff
@@ -0,0 +1,283 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cephfs
+
+import (
+	"context"
+
+	"github.com/ceph/ceph-csi/internal/cephfs/mounter"
+	"github.com/ceph/ceph-csi/internal/cephfs/store"
+	fsutil "github.com/ceph/ceph-csi/internal/cephfs/util"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	mountutil "k8s.io/mount-utils"
+)
+
+type (
+	mountState int
+)
+
+const (
+	msUnknown mountState = iota
+	msNotMounted
+	msMounted
+	msCorrupted
+
+	// ceph-fuse fsType in /proc/<PID>/mountinfo.
+	cephFuseFsType = "fuse.ceph-fuse"
+)
+
+func (ms mountState) String() string {
+	return [...]string{
+		"UNKNOWN",
+		"NOT_MOUNTED",
+		"MOUNTED",
+		"CORRUPTED",
+	}[int(ms)]
+}
+
+func getMountState(path string) (mountState, error) {
+	isMnt, err := util.IsMountPoint(path)
+	if err != nil {
+		if util.IsCorruptedMountError(err) {
+			return msCorrupted, nil
+		}
+
+		return msUnknown, err
+	}
+
+	if isMnt {
+		return msMounted, nil
+	}
+
+	return msNotMounted, nil
+}
+
+func findMountinfo(mountpoint string, mis []mountutil.MountInfo) int {
+	for i := range mis {
+		if mis[i].MountPoint == mountpoint {
+			return i
+		}
+	}
+
+	return -1
+}
+
+// Ensures that given mountpoint is of specified fstype.
+// Returns true if fstype matches, or if no such mountpoint exists.
+func validateFsType(mountpoint, fsType string, mis []mountutil.MountInfo) bool {
+	if idx := findMountinfo(mountpoint, mis); idx > 0 {
+		mi := mis[idx]
+
+		if mi.FsType != fsType {
+			return false
+		}
+	}
+
+	return true
+}
+
+// tryRestoreFuseMountsInNodePublish tries to restore staging and publish
+// volume moutpoints inside the NodePublishVolume call.
+//
+// Restoration is performed in following steps:
+// 1. Detection: staging target path must be a working mountpoint, and target
+//    path must not be a corrupted mountpoint (see getMountState()). If either
+//    of those checks fail, mount recovery is performed.
+// 2. Recovery preconditions:
+//    * volume secrets are present,
+//    * NodeStageMountinfo is present for this volume,
+//    * if staging target path and target path are mountpoints, they must be
+//      managed by ceph-fuse,
+//    * VolumeOptions.Mounter must evaluate to "fuse".
+// 3. Recovery:
+//    * staging target path is unmounted and mounted again using ceph-fuse,
+//    * target path is only unmounted; NodePublishVolume is then expected to
+//      continue normally.
```

Comment:
@gman0 if we `also record nodeStageSecret` information in this record path for a volume, cant we forget about the requirement of filling `nodePublishSecret` and consuming it? because the recovery in any of the above workflow `contact` or `get` a reference about staging information from this record. Please correct if I missed anything here.

Reply:
@humblec I remember there were discussions about safety implications of storing secrets on the node in the past, and I'm not aware there have been any changes in this regard, but please do let me know if you know more. Not having to push the secrets into `NodePublishVolume` calls would be a lot nicer, I completely agree, but I'm not sure we have any safe way to do that.

Even though what ceph-csi is currently doing with the secrets file is not completely bulletproof either, storing the secrets in mountinfo would be a lot worse (it's a `hostPath` volume) I think.

Reply:
The main thought process behind getting rid of nodepublish was that, its a completely new input for the driver at this stage, so I was thinking some way to get rid of it and may be stick to what is known till now for the driver/end user. Actually storing the secret information in the node may not be that bad as even today if someone has access to the host volumes, there is security issue.

Reply:
Could someone with root access to the node (or at least to the node plugin containers) be able to somehow mine the password out of ceph-fuse process memory? What I'm getting at is if it's reasonably easy to retrieve the cephx key using this channel, I'd agree with you that it might not make that much difference if we stored it in a file.

> its a completely new input for the driver at this stage,

Yes, it is very ugly, and I too wish we could get rid of it.

Reply:
Is there someone from Ceph we could ping to verify this claim?

Reply:
I tried asking on ceph-users mailing list:

https://lists.ceph.io/hyperkitty/list/ceph-users@ceph.io/thread/TMLSACFXFBDTUCSVRHQGDORRCJWE24AQ/

Reply:
Thanks,! at present we have one reply which talks about core dumping  as you noted other than the root access.

Reply:
@humblec I've opened https://github.com/ceph/ceph-csi/issues/2924 and I think I can move forward on this. If anyone thinks otherwise please do let me know.

Reply:
Sure, lets get this in this form and continue discussions of possible options in the mentioned issue link.

Reply:
Done, mount info now stores secrets from `NodeStageVolumeRequest` in a file (0600, owned by root).